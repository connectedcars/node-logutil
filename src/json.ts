// https://github.com/microsoft/TypeScript/issues/1897
export type Json = null | boolean | number | string | { [prop: string]: Json } | Json[]
export type JavaScriptValue =
  | Json
  | Buffer
  | Date
  | undefined
  | JavaScriptValue[]
  | { [prop: string]: JavaScriptValue | undefined }
  | Error

export function objectToJson(jsValue: JavaScriptValue): Json {
  const seen: JavaScriptValue[] = []
  return _objectToJson(jsValue, seen, 1)
}

function _objectToJson(jsValue: JavaScriptValue, seen: JavaScriptValue[], level: number): Json {
  if (level >= 10) {
    return `(MaxDepth:${JSON.stringify(_objectToJson(jsValue, seen, Number.MIN_SAFE_INTEGER))})`
  }
  if (jsValue === null) {
    return null
  }
  const type = typeof jsValue
  if (type === 'boolean') {
    return jsValue as boolean
  }
  if (type === 'number') {
    if (isNaN(jsValue as number)) {
      return '(NaN)'
    }
    if (!isFinite(jsValue as number)) {
      return '(Infinity)'
    }
    return jsValue as number
  }
  if (type === 'bigint') {
    return `BigInt(${jsValue?.toString()})`
  }
  if (type === 'string') {
    return (jsValue as string).replace(/\n/g, '\\n')
  }
  if (type === 'object') {
    if (Array.isArray(jsValue)) {
      seen.push(jsValue)
      if (jsValue.length === 0) {
        return jsValue as Json[]
      }
      const values: Json[] = []
      for (const value of jsValue) {
        if (seen.indexOf(value) > -1) {
          values.push('(Circular:StrippedOut)')
        } else {
          values.push(_objectToJson(value, seen, level + 1))
        }
      }
      return values
    }
    if (jsValue instanceof Date) {
      return `Date(${jsValue.toISOString()})`
    }
    if (Buffer.isBuffer(jsValue)) {
      return `Buffer(${jsValue.toString('hex').toUpperCase()})`
    }
    if (jsValue instanceof Error) {
      const stack = typeof jsValue.stack === 'string' ? parseStack(jsValue.stack, jsValue.constructor.name) : []

      let cause = {}
      const _cause = jsValue.cause
      if ('cause' in jsValue && _cause !== undefined && isJavaScriptValue(_cause)) {
        seen.push(jsValue)
        cause = { cause: _objectToJson(_cause, seen, level + 1) }
      }
      return { type: jsValue.constructor.name, message: jsValue.message, stack, ...cause }
    }
    // Object
    const keys = Object.keys(jsValue as { [key: string]: Json | undefined })
    if (keys.length === 0) {
      return jsValue as { [key: string]: Json }
    }
    seen.push(jsValue)
    const obj: { [key: string]: Json } = {}
    for (const key of keys) {
      const value = (jsValue as { [key: string]: Json | undefined })[key]
      if (typeof value === 'undefined') {
        obj[key] = '(undefined)'
      } else {
        if (seen.indexOf(value) > -1) {
          obj[key] = '(Circular:StrippedOut)'
        } else {
          obj[key] = _objectToJson(value, seen, level + 1)
        }
      }
    }
    return obj
  } else {
    throw new Error(`Unknown JavaScript type: ${type}: ${jsValue}`)
  }
}

// Parse a node stack trace to an array of strings,
// fallback to the original stack if the format is unknown
function parseStack(stack: string, className: string, maxDepth = 5): string | string[] {
  if (!stack.startsWith(`${className}: `)) {
    return stack
  }
  const stackLines = []
  for (const line of stack.split('\n')) {
    if (line.startsWith('    at ')) {
      stackLines.push(line.substring(7))
    } else if (line.startsWith(`${className}: `)) {
      // skip
    } else {
      return stack
    }
  }
  return stackLines.slice(0, maxDepth)
}

function isJavaScriptValue(x: unknown): x is JavaScriptValue {
  if (x === null || x === undefined) {
    return true
  }
  const types = ['boolean', 'number', 'bigint', 'string', 'object']
  if (types.includes(typeof x)) {
    return true
  }
  return false
}
