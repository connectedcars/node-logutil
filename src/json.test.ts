/* eslint-disable no-console */
import { JavaScriptValue, objectToJson } from './json'

describe('src/json', () => {
  const sample: any = {
    string: 'string',
    int: 1,
    float: 1.1,
    null: null,
    nan: NaN,
    undefined: undefined,
    date: new Date('2021-01-09T18:11:34.309Z'),
    false: false,
    true: true,
    infinity: 1000 / 0,
    intArray: [1, 2, 3, 4],
    object: {
      stringKey: 'string',
      intKey: 1
    },
    multiLineString: 'hello\nhello',
    emptyArray: [],
    buffer: Buffer.from('DEADBEEF', 'hex'),
    bigInt: 1n,
    deep: {
      deep: {
        deep: {
          deep: {
            deep: {
              deep: {
                deep: {
                  deep: {
                    deep: {
                      deep: {
                        deep: {}
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
  sample.mySelf = sample

  it('should generate JSON where all exceptions are handled', () => {
    const output = objectToJson(sample, { maxObjectSize: 50 })
    expect(output).toMatchSnapshot()

    // Try to convert to JSON and make sure nothing is lost
    const json = JSON.stringify(output, null, 2)
    const sample2 = JSON.parse(json)
    expect(sample2).toEqual(output)
  })

  it('should handle class', () => {
    const output = objectToJson(new SomeTestClass() as unknown as JavaScriptValue)
    expect(output).toEqual({
      __constructorName: 'SomeTestClass',
      someProperty: 'someProperty',
      someField: 'someField'
    })
  })

  describe('Error', () => {
    class HalClientError extends Error {
      public constructor(message?: string) {
        super(message)
        this.name = this.constructor.name
      }
    }
    class ParserError extends HalClientError {}

    it('should handle Error', () => {
      const output = objectToJson(new Error('test error')) as unknown as ErrorJson
      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeUndefined()
    })

    it('should handle custom error', () => {
      const output = objectToJson(new HalClientError('test error')) as unknown as ErrorJson
      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('HalClientError')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeUndefined()
    })

    it('should handle Error with cause', () => {
      const output = objectToJson(new Error('test error', { cause: new Error('cause Error') })) as unknown as ErrorJson
      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeDefined()
      expect(output.cause?.message).toBe('cause Error')
      expect(output.cause?.__constructorName).toBe('Error')
      expect(Array.isArray(output.cause?.stack)).toBe(true)
      expect(output.cause?.stack.length).toBeGreaterThan(0)
    })

    it('should handle Error with custom cause', () => {
      const output = objectToJson(
        new Error('test error', { cause: new ParserError('cause Error') })
      ) as unknown as ErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeDefined()
      expect(output.cause?.message).toBe('cause Error')
      expect(output.cause?.__constructorName).toBe('ParserError')
      expect(Array.isArray(output.cause?.stack)).toBe(true)
      expect(output.cause?.stack.length).toBeGreaterThan(0)
    })

    it('should handle Error with complex cause', () => {
      const output = objectToJson(
        new Error('test error', { cause: { ...sample, error: new ParserError('cause Error') } }),
        { maxObjectSize: 50 }
      ) as unknown as ErrorCauseJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeDefined()
      expect(output.cause?.error?.message).toBe('cause Error')
      expect(output.cause?.error?.__constructorName).toBe('ParserError')
      expect(Array.isArray(output.cause?.error?.stack)).toBe(true)
      expect(output.cause?.error?.stack.length).toBeGreaterThan(0)
    })

    it('should handle Error with circular cause', () => {
      const error = new Error('test error')
      error['cause'] = error
      const output = objectToJson(error) as unknown as ErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.cause).toBeDefined()
      expect(output.cause).toBe('(Circular:StrippedOut)')
    })

    it('handles Error with context', () => {
      const output = objectToJson(
        new ErrorWithContext('test error', {
          cause: new ErrorWithContext('cause Error', { context: { causeContext: 'test' } }),
          context: { foo: 'bar' }
        })
      ) as unknown as ContextErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('ErrorWithContext')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.context).toEqual({ foo: 'bar' })
      expect(output.cause).toBeDefined()
      expect(output.cause?.message).toBe('cause Error')
      expect(output.cause?.__constructorName).toBe('ErrorWithContext')
      expect(Array.isArray(output.cause?.stack)).toBe(true)
      expect(output.cause?.stack.length).toBeGreaterThan(0)
    })

    it('should handle Error with circular context', () => {
      const error = new ErrorWithContext('test error')
      error.context = { error }
      const output = objectToJson(error) as unknown as ContextErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('ErrorWithContext')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.context).toEqual({ error: '(Circular:StrippedOut)' })
    })

    it('should handle Error with undefined context', () => {
      const error = new ErrorWithContext('test error')
      error.context = undefined
      const output = objectToJson(error) as unknown as ContextErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('ErrorWithContext')
      expect(Array.isArray(output.stack)).toBe(true)
      expect(output.stack.length).toBeGreaterThan(0)
      expect(output.context).toBeUndefined()
    })

    it('should handle Error with malformed stack', () => {
      const error = new Error('test error')
      error.stack = 'at foo\nat'
      const output = objectToJson(error) as unknown as ErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(false)
      expect(output.stack).toBe('at foo\nat')
    })

    it('should handle Error with malformed stack at line', () => {
      const error = new Error('test error')
      error.stack = 'Error: \nfoo'
      const output = objectToJson(error) as unknown as ErrorJson

      expect(output.message).toBe('test error')
      expect(output.__constructorName).toBe('Error')
      expect(Array.isArray(output.stack)).toBe(false)
      expect(output.stack).toBe('Error: \nfoo')
    })
  })

  describe('weird cases', () => {
    it('should handle Map', () => {
      const map = new Map()
      map.set('foo', 'bar')
      map.set({ k: 'v' }, () => console.log('hello'))
      map.set(new RegExp('test'), undefined)

      const output = objectToJson(map)
      expect(output).toEqual({ foo: 'bar', '{"k":"v"}': "() => console.log('hello')", 'RegExp(test)': '(undefined)' })
    })

    it('should handle Set', () => {
      const set = new Set()
      set.add('foo')
      set.add(() => console.log('hello'))
      set.add(new RegExp(/\w[test]+/))
      const output = objectToJson(set)
      expect(output).toEqual(['foo', "() => console.log('hello')", 'RegExp(\\w[test]+)'])
    })

    it('should handle weakmap', () => {
      const weakMap = new WeakMap()
      const key = {}
      weakMap.set(key, 'value')
      const output = objectToJson(weakMap)
      expect(output).toEqual('(WeakCollection:strippedOut)')
    })

    it('should handle weakset', () => {
      const weakSet = new WeakSet()
      const key = {}
      weakSet.add(key)
      const output = objectToJson(weakSet)
      expect(output).toEqual('(WeakCollection:strippedOut)')
    })

    it('should handle very long string', () => {
      const longString = 'a'.repeat(1000)
      const output = objectToJson(longString)
      expect(output).toBe('a'.repeat(100) + '...(truncated)')
    })

    it('should handle circular reference in array', () => {
      const arr: any[] = []
      arr.push(arr)
      const output = objectToJson(arr)
      expect(output).toEqual(['(Circular:StrippedOut)'])
    })

    it('should handle too long array', () => {
      const arr = new Array(1000).fill(0)
      const output = objectToJson(arr)
      expect(output).toEqual([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'truncated...'])
    })

    it('should handle empty object', () => {
      const output = objectToJson({})
      expect(output).toEqual({})
    })

    it('should handle too long object', () => {
      const obj: any = {}
      for (let i = 0; i < 1000; i++) {
        obj[i] = i
      }
      const output = objectToJson(obj)
      expect(output).toEqual({ ...Object.fromEntries(Object.entries(obj).slice(0, 10)), 'truncated...': true })
    })
  })
})

interface ErrorJson {
  __constructorName: string
  message: string
  stack: string[]
  cause?: ErrorJson
}

interface ErrorCauseJson {
  __constructorName: string
  message: string
  stack: string[]
  cause?: { error: ErrorJson }
}
interface ContextErrorJson {
  __constructorName: string
  message: string
  stack: string[]
  context: { [key: string]: unknown }
  cause?: ContextErrorJson
}

interface ErrorContext {
  context?: Record<string, unknown>
}

export class ErrorWithContext extends Error {
  public context?: Record<string, unknown>

  public constructor(message: string, options?: ErrorOptions & ErrorContext) {
    super(message, options)
    this.name = this.constructor.name
    this.context = options?.context
  }
}

class SomeTestClass {
  public someProperty = 'someProperty'
  private someField = 'someField'
  public someMethod() {
    return 'someMethod'
  }
}
