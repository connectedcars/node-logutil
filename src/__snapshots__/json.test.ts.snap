// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/json Error should handle Error 1`] = `
Object {
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:67:35)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "Error",
}
`;

exports[`src/json Error should handle Error with cause 1`] = `
Object {
  "cause": Object {
    "message": "cause Error",
    "stack": Array [
      "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:77:68)",
      "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
      "new Promise (<anonymous>)",
      "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
      "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
    ],
    "type": "Error",
  },
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:77:35)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "Error",
}
`;

exports[`src/json Error should handle Error with circular cause 1`] = `
Object {
  "cause": Object {
    "cause": "(Circular:StrippedOut)",
  },
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:94:21)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "Error",
}
`;

exports[`src/json Error should handle Error with complex cause 1`] = `
Object {
  "cause": Object {
    "bigInt": "BigInt(1)",
    "buffer": "Buffer(DEADBEEF)",
    "date": "Date(2021-01-09T18:11:34.309Z)",
    "deep": Object {
      "deep": Object {
        "deep": Object {
          "deep": Object {
            "deep": Object {
              "deep": Object {
                "deep": Object {
                  "deep": Object {
                    "deep": "(MaxDepth:strippedOut:object)",
                  },
                },
              },
            },
          },
        },
      },
    },
    "emptyArray": Array [],
    "error": Object {
      "message": "cause Error",
      "stack": Array [
        "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:88:62)",
        "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
        "new Promise (<anonymous>)",
        "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
        "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
      ],
      "type": "ParserError",
    },
    "false": false,
    "float": 1.1,
    "infinity": "(Infinity)",
    "int": 1,
    "intArray": Array [
      1,
      2,
      3,
      4,
    ],
    "multiLineString": "hello\\\\nhello",
    "mySelf": Object {
      "bigInt": "BigInt(1)",
      "buffer": "Buffer(DEADBEEF)",
      "date": "Date(2021-01-09T18:11:34.309Z)",
      "deep": "(Circular:StrippedOut)",
      "emptyArray": "(Circular:StrippedOut)",
      "false": false,
      "float": 1.1,
      "infinity": "(Infinity)",
      "int": 1,
      "intArray": "(Circular:StrippedOut)",
      "multiLineString": "hello\\\\nhello",
      "mySelf": "(Circular:StrippedOut)",
      "nan": "(NaN)",
      "null": null,
      "object": "(Circular:StrippedOut)",
      "string": "string",
      "true": true,
      "undefined": "(undefined)",
    },
    "nan": "(NaN)",
    "null": null,
    "object": Object {
      "intKey": 1,
      "stringKey": "string",
    },
    "string": "string",
    "true": true,
    "undefined": "(undefined)",
  },
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:88:9)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "Error",
}
`;

exports[`src/json Error should handle Error with custom cause 1`] = `
Object {
  "cause": Object {
    "message": "cause Error",
    "stack": Array [
      "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:82:68)",
      "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
      "new Promise (<anonymous>)",
      "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
      "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
    ],
    "type": "ParserError",
  },
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:82:35)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "Error",
}
`;

exports[`src/json Error should handle custom error 1`] = `
Object {
  "message": "test error",
  "stack": Array [
    "Object.<anonymous> (/Users/ofweb/Code/node-logutil/src/json.test.ts:72:35)",
    "Promise.then.completed (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:293:28)",
    "new Promise (<anonymous>)",
    "callAsyncCircusFn (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/utils.js:226:10)",
    "_callCircusTest (/Users/ofweb/Code/node-logutil/node_modules/jest-circus/build/run.js:248:40)",
  ],
  "type": "HalClientError",
}
`;

exports[`src/json should generate JSON where all exceptions are handled 1`] = `
Object {
  "bigInt": "BigInt(1)",
  "buffer": "Buffer(DEADBEEF)",
  "date": "Date(2021-01-09T18:11:34.309Z)",
  "deep": Object {
    "deep": Object {
      "deep": Object {
        "deep": Object {
          "deep": Object {
            "deep": Object {
              "deep": Object {
                "deep": Object {
                  "deep": Object {
                    "deep": "(MaxDepth:strippedOut:object)",
                  },
                },
              },
            },
          },
        },
      },
    },
  },
  "emptyArray": Array [],
  "false": false,
  "float": 1.1,
  "infinity": "(Infinity)",
  "int": 1,
  "intArray": Array [
    1,
    2,
    3,
    4,
  ],
  "multiLineString": "hello\\\\nhello",
  "mySelf": "(Circular:StrippedOut)",
  "nan": "(NaN)",
  "null": null,
  "object": Object {
    "intKey": 1,
    "stringKey": "string",
  },
  "string": "string",
  "true": true,
  "undefined": "(undefined)",
}
`;
